### PR TITLE
orig dst: add support for port override

### PR DESCRIPTION
**Pull request by briansonnenberg**
Cherrypick of [this commit](https://github.com/istio/envoy/commit/cc9bfc0b956cf2d53bc00675fdb11e4a505d3dd2) with some slight modifications.  Adds a couple test cases and uses WKT.

Co-authored-by: Brian Sonnenberg <bsonnenberg@google.com>
Signed-off-by: Steven Landow <landow@google.com>
Signed-off-by: Brian Sonnenberg <bsonnenberg@google.com>
Signed-off-by: John Howard <howardjohn@google.com>

Commit Message:
Added an option in OriginalDstCluster type to override the original dst port with a defined port.
Additional Description:
Risk Level: Low
Testing: Added tests + all tests passing.
Docs Changes:
Release Notes: TBD
Platform Specific Features:
[Optional Fixes #Issue] [#16503](https://github.com/envoyproxy/envoy/issues/16503)

### CODE REVIEW

File: api/envoy/config/cluster/v3/cluster.proto [hunk 1]
```diff
@@ -546,6 +546,8 @@ message Cluster {
     // The http header to override destination address if :ref:`use_http_header <envoy_v3_api_field_config.cluster.v3.Cluster.OriginalDstLbConfig.use_http_header>`.
     // is set to true. If the value is empty, :ref:`x-envoy-original-dst-host <config_http_conn_man_headers_x-envoy-original-dst-host>` will be used.
     string http_header_name = 2;
+
+    uint32 upstream_port_override = 3;
```
Comment (howardjohn) [inline api/envoy/config/cluster/v3/cluster.proto:550]: This could probably use a bit of documentation describing what it does

File: source/common/upstream/original_dst_cluster.cc [hunk 2]
```diff
@@ -39,6 +39,9 @@ HostConstSharedPtr OriginalDstCluster::LoadBalancer::chooseHost(LoadBalancerCont
         dst_host = connection->connectionInfoProvider().localAddress();
       }
     }
+    if (dst_host && port_override_ > 0) {
+      dst_host = Network::Utility::getAddressWithPort(*dst_host.get(), port_override_);
```
Comment (howardjohn) [inline source/common/upstream/original_dst_cluster.cc:43]: if `getAddressWithPort` returns `null` then dst_host will be set to null as well I think - maybe want to only override if its non-null?

**Review by howardjohn — DISMISSED** [review]
Looks great!

File: source/common/upstream/original_dst_cluster.cc (hunk 2 above)
Comment (briansonnenberg) [reply source/common/upstream/original_dst_cluster.cc:43]: If getAddressWithPort returns null then something has gone seriously wrong and I think we should fall through to line 86.

File: source/common/network/utility.cc [hunk 3]
```diff
@@ -328,8 +328,9 @@ Address::InstanceConstSharedPtr Utility::getAddressWithPort(const Address::Insta
     return std::make_shared<Address::Ipv4Instance>(address.ip()->addressAsString(), port);
   case Address::IpVersion::v6:
     return std::make_shared<Address::Ipv6Instance>(address.ip()->addressAsString(), port);
+  default:
+    return nullptr;
   }
-  PANIC("not handled");
 }
```
Comment (mattklein123) [inline source/common/network/utility.cc:334]: Why was this changed? What case causes this to be hit now?

File: api/envoy/config/cluster/v3/cluster.proto [hunk 4]
```diff
@@ -546,6 +546,10 @@ message Cluster {
     // The http header to override destination address if :ref:`use_http_header <envoy_v3_api_field_config.cluster.v3.Cluster.OriginalDstLbConfig.use_http_header>`.
     // is set to true. If the value is empty, :ref:`x-envoy-original-dst-host <config_http_conn_man_headers_x-envoy-original-dst-host>` will be used.
     string http_header_name = 2;
+
+    // The port to override for the original dst address. This port
+    // will take precedence over filter state and header override ports
+    uint32 upstream_port_override = 3;
```
Comment (mattklein123) [inline api/envoy/config/cluster/v3/cluster.proto:552]: This should be a WKT to avoid relying on 0 being not set.

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks. This makes sense at a high level. A few questions to get started.

/wait

File: source/common/network/utility.cc (hunk 3 above)
Comment (briansonnenberg) [reply source/common/network/utility.cc:334]: There was [some concern](https://github.com/howardjohn/envoy/commit/53a23d5a3df705338c776eb6c500d793218a0367#r74719909) that we might panic in the original PR.  I figured it's better to allow the callers to handle a nullptr than to just give up.  Fine with reverting this if we don't care.

Comment (mattklein123) [reply source/common/network/utility.cc:334]: Let's just revert for now.

File: source/common/upstream/original_dst_cluster.h [hunk 5]
```diff
@@ -77,10 +78,14 @@ class OriginalDstCluster : public ClusterImplBase {
     const std::shared_ptr<OriginalDstCluster> parent_;
     // The optional original host provider that extracts the address from HTTP header map.
     const absl::optional<Http::LowerCaseString>& http_header_name_;
+    const bool use_port_override_;
+    const uint32_t port_override_;
```
Comment (mattklein123) [inline source/common/upstream/original_dst_cluster.h:82]: use absl::optional here to avoid 2 variables.

**Review by mattklein123 — CHANGES_REQUESTED** [review]
LGTM with small comment.

/wait

File: api/envoy/config/cluster/v3/cluster.proto [hunk 6]
```diff
@@ -546,6 +546,10 @@ message Cluster {
     // The http header to override destination address if :ref:`use_http_header <envoy_v3_api_field_config.cluster.v3.Cluster.OriginalDstLbConfig.use_http_header>`.
     // is set to true. If the value is empty, :ref:`x-envoy-original-dst-host <config_http_conn_man_headers_x-envoy-original-dst-host>` will be used.
     string http_header_name = 2;
+
+    // The port to override for the original dst address. This port
+    // will take precedence over filter state and header override ports
+    google.protobuf.UInt32Value upstream_port_override = 3;
```
Comment (mattklein123) [inline api/envoy/config/cluster/v3/cluster.proto:552]: Please add a release note for this.

File: source/common/upstream/original_dst_cluster.h [hunk 7]
```diff
@@ -128,6 +130,7 @@ class OriginalDstCluster : public ClusterImplBase {
   absl::Mutex host_map_lock_;
   HostMapConstSharedPtr host_map_ ABSL_GUARDED_BY(host_map_lock_);
   absl::optional<Http::LowerCaseString> http_header_name_;
+  absl::optional<uint32_t> port_override_{};
```
Comment (mattklein123) [inline source/common/upstream/original_dst_cluster.h:133]: nit

```suggestion
  absl::optional<uint32_t> port_override_;
```

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks LGTM with small remaining comments.

/wait

File: api/envoy/config/cluster/v3/cluster.proto (hunk 6 above)
Comment (soulxu) [inline api/envoy/config/cluster/v3/cluster.proto:552]: We can add validation `[(validate.rules).uint32 = {lte: 65535}]` just like https://github.com/envoyproxy/envoy/blob/667f0600743104857d59f56da4ab6aeea57fc60e/api/envoy/config/core/v3/address.proto#L78

**Review by soulxu — CHANGES_REQUESTED** [review]

**Review by mattklein123 — CHANGES_REQUESTED** [review]
LGTM modulo remaining PGV comment, thank you.

/wait

**Comment by soulxu** [comment]
thanks for the update! LGTM

**Review by soulxu — DISMISSED** [review]

**Review by mattklein123 — DISMISSED** [review]
Thanks!

**Comment by mattklein123** [comment]
Oops sorry looks like you need a main merge.

/wait

**Review by mattklein123 — APPROVED** [review]
Thanks!
